### ISSUE TITLE
Racon polishing removes scaffolds

### ISSUE THREAD

**Issue by SorenKarst** [isovic/racon#74]
We are working with microbial metagenomics, where we use nanopore data for de novo assembly with miniasm followed by polishing with racon. 

After the last racon update (1.2.0) we can see that some scaffolds are removed when we perform racon polishing. I just wondered if this is intentional and if it is what the criteria are for this filtering (i.e. mapping coverage or quality)?

regards
Søren

Ps. Thank you for an amazing tool.

**Comment by rvaser** [comment]
Hi Søren,
thanks for the kind words. Are you using anything in between miniasm and racon?

If not, some scaffolds might not be corrected due to low coverage (or maybe something went wrong), you can obtain them by using the option:
```bash
    -u, --include-unpolished
         output unpolished target sequences
````

Best regards,
Robert

**Comment by SorenKarst** [comment]
Hi Robert,

Thank you for the quick reply.

We do perform some custom formatting between miniasm and racon to streamline headers and filter short scaffolds. The nanopore data is with quality values. The dataset is 14.5 Gbp and we expect some of the organisms to be in low relative abundance (< 1 x coverage) and a lot of complicated repeats, so I would expect some of the miniasm assemblies to be poorly supported.

The following is the code we use:

minimap2-t 40 -x ava-ont all_nptrim.fq \
all_nptrim.fq > read_minimap.paf

miniasm -f all_nptrim.fq read_minimap.paf \
> assembly_miniasm.gfa

awk '/^S/{print ">"$2"\n"$3}' assembly_miniasm.gfa |\ # gfa to fasta
awk '/^>/{print ">" ++i; next}{print}' | \ # Names scaffolds 1 to n.
awk -v seq_minlen="4000" ' # Removes scaffolds below 4000 bp
  !/^>/ { next } { getline seq } length(seq) >= seq_minlen { print $0 "\n" seq }' - \
> metagenome_assembly/assembly_miniasm.fa

minimap2 -x map-ont -t 40 assembly_miniasm.fa \
all_nptrim.fq > assembly_minimap.paf

racon -t 40 all_nptrim.fq \
assembly_minimap.paf assembly_miniasm.fa \
> assembly_racon.fa
sed -i '/^>/s/ .*//' assembly_racon.fa # removes racon meta

best regards
Søren

**Comment by rvaser** [comment]
Hi Søren,
the filtering process in racon retains only one mapping (with error rate less than 30%) per read in polishing mode, parts of reads of length 500 which have avg quality < 10 are not used for polishing (you might want to check the average base quality of the whole read set and set the parameter accordingly) and each window of a target sequence is not polished if there are less than 3 reads mapped to it. What percentage of scaffolds are dropped when you run the above pipeline?

Best regards,
Robert

**Comment by SorenKarst** [comment]
Hi again,

It's only 14 scaffolds out of 3216, and they are small compared to the rest of the data (7 - 35 kbp). It makes perfect sense that these scaffolds are dropped due to low coverage/quality. If I understand you correctly the -u command will retain these scaffolds in the assembly?

regards
Søren

**Comment by rvaser** [comment]
Exactly, the -u option will output the unpolished scaffolds with the polished ones. Outputted sequences in headers have three SAM-like tags which describe their length, approximate number of reads used in polishing and percentage of polished windows.

You can also try the -f option which disables the one mapping per read rule (primarily used for self error correction of reads) which increases the execution time.

Best regards,
Robert

**Comment by SorenKarst** [comment]
It worked liked a charm.

Thanks a lot for the help and the awesome tool.

best regards,
Søren

**Comment by Gsmith535** [comment]
I know this issue was closed a long time ago now, but I ran racon v1.3.1 with -u and still noticed scaffolds/contigs being removed; in fact, I saw a consistent single seqeunce removal while looping through iterations after the first. This occurred for two different communities with two assemblers.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/smith10284/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/smith10284/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | Community A | Community A | Community B | Community B
-- | -- | -- | -- | --
  | assembler X | assembler Y | assembler X | assembler Y
racon iterations | # sequences | # sequences | # sequences | # sequences
0 | 4187 | 1487 | 6951 | 2301
1 | 4187 | 1487 | 6951 | 2301
2 | 4186 | 1486 | 6950 | 2300
3 | 4185 | 1485 | 6949 | 2299
4 | 4184 | 1484 | 6948 | 2298
5 | 4183 | 1483 | 6947 | 2297
6 | 4182 | 1482 | 6946 | 2296
7 | 4181 | 1481 | 6945 | 2295
8 | 4180 | 1480 | 6944 | 2294
9 | 4179 | 1479 | 6943 | 2293
10 | 4178 | 1478 | 6942 | 2292



</body>

</html>



According to the headers, the number of contigs to which no reads were used for polishing tended to increase but sometimes decreased. The overall total number of reads used for polishing tended to increase but changed <1.2% overall after the first iteration.

I assumed the slight decrease in sequence count meant that similar sequences were collapsed into a consensus, but I did not realize that I lost a large and apparently unique scaffold, as well as the consistency of the decreases until reading this post. Is this something that -f might have prevented? I did not understand the function of this flag when I originally ran the program.

I performed no steps between minimap and racon. In between iterations I adjusted the headers to include the number of times racon was run, but that was the only additional step. 

simplified commands:
```
minimap2 -t 6 -c -x map-ont asm.fa reads.qc.fq > reads_map2_asm.paf
racon -t 6 -u  reads.qc.fq reads_map2_asm.paf asm.fa > asm_rac.fa
sed 's|>|>rac1_|' asm_rac.fa > asm_rac1.fa # adding racon run identifier to headers 
for n in 1:10; do 
minimap2 -t 6 -c -x map-ont asm_"$n".fa reads.qc.fq > reads_map2_asm_rac"$n".paf
racon -t 6 -u  reads.qc.fq reads_map2_asm_rac"$n".paf asm_rac"$n".fa > asm_rac.fa
sed "s|rac"$n"|rac"$n+1"|" asm_rac.fa > asm_rac"$n+1".fa # update racon run identifiers in headers
done
```

**Comment by rvaser** [comment]
Hello,
could you please verify this behavior with the latest version `1.5.0`?

Best regards,
Robert
